crates/matrix-sdk/src/latest_events/mod.rs

```diff
@@ -0,0 +1,861 @@
+    /// fetch the [`LatestEvent`] for a particular thread in this room.
+    ///
+    /// It returns `None` if the room or the thread doesn't exist.
+    ///
+    /// Note that a read lock guard is returned if the [`LatestEvent`] exists.
+    /// One doesn't want to keep this lock alive for too long, otherwise the
+    /// whole [`LatestEvents`] API is blocked. Use [`LatestEvent::get`] or
+    /// [`LatestEvent::subscribe`] as quick as possible, and then drop this
+    /// lock guard.
+    pub async fn for_thread(
+        &self,
+        room_id: &RoomId,
+        thread_id: &EventId,
+    ) -> Option<RwLockReadGuard<'_, LatestEvent>> {
+        self.state.registered_rooms.for_thread(room_id, thread_id).await
+    }
+
+    /// Forget a room.
+    ///
+    /// It means that [`LatestEvents`] will stop listening to updates for the
+    /// `LatestEvent`s of the room and all its threads.
+    ///
+    /// If [`LatestEvents`] is not listening for `room_id`, nothing happens.
+    pub async fn forget_room(&self, room_id: &RoomId) {
+        self.state.registered_rooms.forget_room(room_id).await;
+    }
+
+    /// Forget a thread.
+    ///
+    /// It means that [`LatestEvents`] will stop listening to updates for the
+    /// `LatestEvent` of the thread.
+    ///
+    /// If [`LatestEvents`] is not listening for `room_id` or `thread_id`,
+    /// nothing happens.
+    pub async fn forget_thread(&self, room_id: &RoomId, thread_id: &EventId) {
+        self.state.registered_rooms.forget_thread(room_id, thread_id).await;
+    }
+}
+
+#[derive(Debug)]
+struct RegisteredRooms {
+    /// All the registered [`RoomLatestEvents`].
+    rooms: RwLock<HashMap<OwnedRoomId, RoomLatestEvents>>,
+
+    /// The sender part of the channel about room registration.
+    ///
+    /// When a room is registered (with [`LatestEvents::for_room`] or
+    /// [`LatestEvents::for_thread`]) or unregistered (with
+    /// [`LatestEvents::forget_room`] or [`LatestEvents::forget_thread`]), a
+    /// room registration message is passed on this channel.
+    ///
+    /// The receiver part of the channel is in the
+    /// [`listen_to_event_cache_and_send_queue_updates_task`].
+    room_registration_sender: mpsc::Sender<RoomRegistration>,
+}
+
+impl RegisteredRooms {
+    fn new(room_registration_sender: mpsc::Sender<RoomRegistration>) -> Self {
+        Self { rooms: RwLock::new(HashMap::default()), room_registration_sender }
+    }
+
+    /// Get a read lock guard to a [`RoomLatestEvents`] given a room ID and an
+    /// optional thread ID.
+    ///
+    /// The [`RoomLatestEvents`], and the associated [`LatestEvent`], will be
+    /// created if missing. It means that write lock is taken if necessary, but
+    /// it's always downgraded to a read lock at the end.
+    async fn room_latest_event(
+        &self,
+        room_id: &RoomId,
+        thread_id: Option<&EventId>,
+    ) -> Option<RwLockReadGuard<'_, RoomLatestEvents>> {
+        match thread_id {
```

Comment:
It looks like a lot of this code is duplicated between the 2 match branches. Can we if let the thread in the second step and reduce the indentation here?

Reply:
Sadly no, because in one case we take a write-lock immediately, another time, we take a read-lock then a write-lock, it's a bit different.